llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -325,23 +325,41 @@ bool RepoPruning::runOnModule(Module &M) {
   }
 
   // Enable this pass support for alias.
-  for (Module::alias_iterator I = M.alias_begin(), E = M.alias_end(); I != E;) {
-    Module::alias_iterator CurI = I;
-    ++I;
+  // First, check whether a GlobalAlias references a pruned GO. If yes, record
+  // it. Second, replace all recorded aliases by the declaration. These two
+  // steps can't be merged since the second step could change the base object of
+  // the alias. For example:
+  // Assuming: 1) the aliasee of alias 'A' is alias 'B',
+  //           2) the aliasee of alias 'B' is global object 'C',
+  //           3) the global object 'C' is pruned.
+  // If merging two steps, the order in the alias list is 'B' and then 'A'.
+  // Since the global object 'C' is prund, the alias 'B' is changed from
+  // defination to declaration. Then, since 'B' is changed to declaration, the
+  // base object of the alias 'A' will be changed from 'C' to 'B'. This will
+  // result in the backend error.
+  // Step 1. Record aliases.
+  SmallVector<GlobalAlias *, 1> Aliases;
+  for (Module::alias_iterator I = M.alias_begin(), E = M.alias_end(); I != E;
+       ++I) {
     // TODO: An Alias doesn't have an aliasee, for example:
     // @v = alias i32, inttoptr(i32 42 to i32*)
-    if (auto *Aliasee = CurI->getBaseObject()) {
+    if (auto *Aliasee = I->getBaseObject()) {
       if (wasPruned(*Aliasee)) {
-        Type *Ty = CurI->getValueType();
-        // Delete the alias from Module if its aliasee is pruned.
-        CurI->removeFromParent();
-        // Replace the alias by a declaration.
-        CurI->replaceAllUsesWith(createDeclararion(M, Ty, CurI));
-        delete &*CurI;
+        LLVM_DEBUG(dbgs() << "Alias: " << I->getName() << ", its base object: "
+                          << Aliasee->getName() << " is pruned! \n");
+        Aliases.push_back(&*I);
         ++NumAliases;
       }
     }
   }
+  // 2. Resolve aliases.
+  for (unsigned I = 0, E = Aliases.size(); I < E; ++I) {
+    Type *Ty = Aliases[I]->getValueType();
+    // Replace the alias by a declaration.
+    Aliases[I]->replaceAllUsesWith(createDeclararion(M, Ty, Aliases[I]));
+    // Delete the alias from Module if its aliasee is pruned.
```

Comment:
This comment is slightly confusing in the new code. I think that the list `Aliases` is of aliases to pruned objects so saying "if its aliasee is pruned" at this point.

Reply:
Fixed, thanks.